cmake-modules/add_nlohmann_json.cmake

```diff
@@ -0,0 +1,27 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# SPDX-License-Identifier: MIT
+#
+# Azure Core can expose the Nlohmann Json header by fetching a release version and including the header to the include path.
+# 
+# Azure Core will try to find the json header first. This is for an application that is already using Nlohmann header.
+# When it is not found it, Azure Core will fetch it.
+
+# Storage requires 3.6.0 version for using `contains` feature
+# Using QUIET to avoid warning message if lib is not found
+find_package(nlohmann_json 3.6.0 EXACT QUIET)
```

Comment:
For future reference, we should not have downgraded from 3.8.0 to 3.6.0, especially in this PR which is replacing the existing checked-in source (which pointed to 3.8.0). Going forward, let's update to newest versions of the library with any bug fixes.

Reply:
I still didn't get why we need `EXACT` keyword here.

> Storage requires 3.6.0 version

Actually we require version >= 3.6.0, not exact 3.6.0. Because usually when some library or software adds a new feature in some version, they keep this feature in all the following versions. It's not like they added this feature in 3.6.0 then removed in 3.6.1.

Reply:
https://github.com/Azure/azure-sdk-for-cpp/issues/1027

Reply:
EXACT is used to prevent nlohmann/json to add a breaking change in the future that would impact us.

There is a good thing and a bad thing about EXCAT

Good: No breaking changes in the future.
Bad: Azure SDK won't work if the end-user is already using a different version. Unless they remove this EXACT keyword and validate if they can build without issues.

In theory, if we say "we support 3.6 and greater, we would need to test all versions after 3.6 for every release we do to ensure this is true. Otherwise, we are trying to be in a safe zone where checking if the newer version works is something that the end-user needs to try/verify. 

Maybe we can add a build option to disable the EXACT 3rd parties

Reply:
Then how about libcurl https://github.com/Azure/azure-sdk-for-cpp/blob/c9c12e27aabe5ce971de33eef22db6d0fc97f0cb/sdk/core/azure-core/CMakeLists.txt#L20? Why don't we try to avoid breaking changes in libcurl?

Reply:
libcurl provides a compatibility configuration that cmake can use:
https://github.com/curl/curl/blob/f21cc62832a9d5c521d10da9177b075bfed43f2f/CMakeLists.txt#L1514-L1519

See: https://cmake.org/cmake/help/latest/module/CMakePackageConfigHelpers.html

It looks like nlohmann json has a similar compatibility declaration, though they don't explicitly use the built-in `SameMajorVersion` cmake literal:
https://github.com/nlohmann/json/blob/fd7a9f600712b2724463e9f7f703878ade676d6e/cmake/nlohmann_jsonConfigVersion.cmake.in#L1

We discussed this a bit today. Though it is useful to do a case-by-case analysis on this topic, based on factors such as pervasiveness, versioning strategy of each dependency or library we use, etc. it is fine to use `config` as the default as a starting point.

Hopefully, either choice here won't have much negative impact to the end user, so it isn't extremely critical, **but we can update this JSON library to use `config`.** This relies on the following assumptions:
1) The library makes an attempt to avoid breaking changes or reduces them as much as possible (looking at its history or stance).
2) When it does make such changes, it follows appropriate semver semantics, and revs the major version.
3) It declares in its cmake package config file that either `AnyNewerVersion` or `SameMajorVersion` (or equivalent) is considered compatible.

Reply:
I don't understand why you guys make so big deal out of it. To me, `find_package(nlohmann_json REQUIRED)` is enough and fits most our customers' needs. But anyway, you're the owner of this code and you decide what to do.

Reply:
Regardless of code ownership, this conversation has been great since it encourages a deeper understanding of how the various cmake features work, the nuances of our design choices, and the impact on the end user.